docs/rfc/block-registration.md

```diff
@@ -0,0 +1,517 @@
+			"selector": "img",
+			"attribute": "src"
+		},
+		"author": {
+			"type": "string",
+			"source": "html",
+			"selector": ".book-author"
+		}
+	}
+}
+```
+
+Attributes provide the structured data needs of a block. They can exist in different forms when they are serialized, but they are declared together under a common interface.
+
+See the [the attributes documentation](/docs/designers-developers/developers/block-api/block-attributes.md) for more details.
+
+### Style Variations
+
+* Type: `array`
+* Optional
+* Localized: Yes (`label`)
+* Property: `styleVariations`
+* Alias: `styles`
+
+```json
+{ 
+	"styleVariations": [ 
+		{ "name": "default", "label": "Default", "isDefault": true }, 
+		{ "name": "other", "label": "Other" }
+	]
+}
+```
+
+Block styles can be used to provide alternative styles to block. It works by adding a class name to the block's wrapper. Using CSS, a theme developer can target the class name for the style variation if it is selected.
+
+Plugins and Themes can also register [custom block style](/docs/designers-developers/developers/filters/block-filters.md#block-style-variations) for existing blocks.
+
+### Editor Script
+
+* Type: `string|object` ([WPDefinedAsset](#WPDefinedAsset))
+* Optional
+* Localized: No
+* Property: `editorScript`
+
+```json
+{ "editorScript": "build/editor.js" }
+```
+
+Block type editor script definition. It will only be enqueued in the context of the editor.
+
+### Script
+
+* Type: `string|object` ([WPDefinedAsset](#WPDefinedAsset))
+* Optional
+* Localized: No
+* Property: `script`
+
+```json
+{
+	"script": {
+		"handle": "my-plugin-notice",
+		"file": "index.js",
+		"dependencies": [ "wp-element", "wp-components" ],
+		"version": "3.0.0"
+	}
+}
+```
+
+Block type frontend script definition. It will be enqueued both in the editor and when viewing the content on the front of the site.
+
+### Editor Style
+
+* Type: `string|object` ([WPDefinedAsset](#WPDefinedAsset))
+* Optional
+* Localized: No
+* Property: `editorStyle`
+
+```json
+{
+	"editorStyle": {
+		"handle": "my-plugin-notice-editor",
+		"file": "editor.css",
+		"dependencies": [ "wp-edit-blocks" ]
+	}
+}
+```
+
+Block type editor style definition. It will only be enqueued in the context of the editor.
+
+### Style
+
+* Type: `string|object` ([WPDefinedAsset](#WPDefinedAsset))
+* Optional
+* Localized: No
+* Property: `style`
+
+```json
+{ "style": "build/style.css" }
+```
+
+Block type frontend style definition. It will be enqueued both in the editor and when viewing the content on the front of the site.
+
+### Render Callback
+
+* Type: `string` ([WPDefinedPropertyFile](#WPDefinedPropertyFile))
+* Optional
+* Localized: No
+* Property: `renderCallback`
+
+```json
+{ "renderCallback": "my-render-callback.php" }
+```
+
+This is a pointer to a php file returning a render callback php function. The render callback is function called when the block is rendered on the frontend. It's used to generate the frontend markup dynamically.
```

Comment:
Yeah, this is a little confusing. It sounds like the expectation is that you would have a file that returns a callable…? Something like this?

```
<?php

return function ( $attributes ) {
    return '<div> My block markup </div>';
};
```

I'm not sure there's any precedent for this in Core (I'm not even sure when it became possible to return closures; is that available in PHP 5.6?) and it honestly feels like a strange solution overall.

I understand the intent here is for a unified, centralised API but this almost feels like a case where the better approach would be for the central block definition format to be agnostic to the existence of render callbacks, and have it so you actually _bind your `render_callback` to the block_ through a WordPress core filter or something like that. In other words, the server side rendering stops being an intrinsic property of the block, and is instead an implementation that is specific to the rendering engine, whether that's WordPress or another framework.

```
<?php
// Pseudo-code; don't take too seriously.

$block_html = apply_filters( 'wp_{$block_name}_render_callback', $block_html, $attributes, $raw_block_html );
```

This would also better enable custom rendering pipeline for core blocks, and could also open up server side rendering as an option for non-PHP contexts.

Reply:
In PHP callbacks are defined as [such](https://www.php.net/manual/en/language.types.callable.php). 

```
Some functions like call_user_func() or usort() accept user-defined callback functions as a parameter. Callback functions can not only be simple functions, but also object methods, including static class methods.
```

This is extremely confusing as a PHP developer.  Callbacks are functions / methods and not a file. A file like this should be called a template, a template file shouldn't need a callable function in it. 

See how [acf_register_block_type](https://www.advancedcustomfields.com/resources/acf_register_block_type/) docs explain it. 
They list `render_template` and `render_callback` as two different args. I am not sure which one is used if both are defined. It is also much easier to control, if template only ever expects a file and callback a callable function. 

To allow for namespaces and methods on classes to be callbacks, those MUST be defined in php.

Reply:
> I'm not sure there's any precedent for this in Core (I'm not even sure when it became possible to return closures; is that available in PHP 5.6?) and it honestly feels like a strange solution overall.

Core is now PHP 5.6+ so it doesn't seem to be a technical limitation anymore, however, I get your point about the familiarity of the patterns used. In JavaScript world the fact that would be highly expected approach but I see that it might not click here.

> I understand the intent here is for a unified, centralised API but this almost feels like a case where the better approach would be for the central block definition format to be agnostic to the existence of render callbacks, and have it so you actually _bind your `render_callback` to the block_ through a WordPress core filter or something like that.

Well, you made a really good point that it's very specific to WordPress implementation written in PHP. There is a related discussion started by @iandunn where he proposes to render dynamic blocks on the front end in JavaScript instead of PHP. This would fit better to what you describe in your comment.

> ```
> <?php
> // Pseudo-code; don't take too seriously.
> 
> $block_html = apply_filters( 'wp_{$block_name}_render_callback', $block_html, $attributes, $raw_block_html );
> ```
> 
> 
> This would also better enable custom rendering pipeline for core blocks, and could also open up server side rendering as an option for non-PHP contexts.

I wouldn't be opposed personally to removing `renderCallback` from the proposal at all and provide an alternative way to register this `renderCallback` in PHP. There are several ways you could do it, the simplest one could be:

```php
register_block_type(
    'path/to/my/plugin/block.json',
    array( 'render_callback' => 'my_plugin_render_callback' )
);
```

> They list `render_template` and `render_callback` as two different args. I am not sure which one is used if both are defined. It is also much easier to control, if template only ever expects a file and callback a callable function.

The other benefit of keeping it outside of `block.json` is that you could support both ways of registering those callback similar to ACF does.

Reply:
I think that registering the render callback is a runtime implementation. A php file would mean nothing in node or python. This locks the render too php, which isn’t great. 

Render template or better yet, render require, would just require a file. This file could be Javascript / php or whatever manage. It would be up the runtime what to do about that require. In node, if the file ends with .php, it could choose to do nothing with that require. It opens up to Javascript template files.